### PR TITLE
Reconcile PP+MTP daemon path from #382

### PR DESCRIPTION
**Pull request by Kaden-Schutt**
## Summary

Surgical current-master reconciliation of the useful PP+MTP pieces from #382.

This intentionally does **not** import the old `SpecPath` / `GenerateCtx` dispatcher rewrite or the full old branch history. It adds the narrow pieces needed to load a compressed MTP head on the PP output device and run greedy PP+MTP decode through the daemon.

## What changed

- Add `forward_prefill_batch_multi_with_caps` so PP verify can capture output-band hidden rows and elide last-token logits when requested.
- Add PP-safe compressed MTP stepper with output-device MTP head, mirrored token embeddings, and multi-device DeltaNet snapshot/restore.
- Add daemon load/generate wiring for `pp > 1` + explicit compressed `.mtp` head.
- Add `hipfire_runtime::mtp_mirror` tensor clone helpers.

## Scope intentionally left out

- Old #382 `SpecPath` / `GenerateCtx` dispatcher rewrite.
- Old PP GDN tape-shard fast replay. This branch uses full accepted-prefix replay on partial accepts for correctness first.

## Validation

Local:

```bash
cargo check -p hipfire-runtime --example daemon
git diff --check
```

hiptrx rerun after avoiding a midflight collision with another agent:

- Commit: `6d86816f`
- Hardware: `ROCR_VISIBLE_DEVICES=2,3`, 2x Radeon AI PRO R9700 (`gfx1201`), HIP 7.2
- Trunk: `/home/kaden/.hipfire/models/qwen3-35b-a3b.mq4-awq`
  - md5 `edde51ec1dac0f2bd42cff5ef1cb8944`
  - sha256 `1dc1c7964de415e0040a540a4300b9518e11b00c13d99c23f576f2b9fe1e8bca`
- MTP head: `/home/kaden/.hipfire/models/qwen3.6-35b-a3b.moe-mtp-mq4-cvs16384.mtp`
  - md5 `51076bfb5b489832f2f6c4191b9799b0`
  - sha256 `1e11a06d1946e1e5711d6692894e917a61d5f360d4f3508c8372d49e97c912c1`

Load params:

```json
{
  "max_seq": 4096,
  "pp": 2,
  "kv_mode": "q8",
  "mtp_mode": "on",
  "mtp_head": "/home/kaden/.hipfire/models/qwen3.6-35b-a3b.moe-mtp-mq4-cvs16384.mtp",
  "mtp_k": 5,
  "mtp_kv_mode": "q8"
}
```

Result:

- PP split 20/20, `output_device=1`, `peer_access=true`.
- Warmup: `spec_path:"pp-mtp"`, coherent `warmup<|im_end|>`.
- Smoke prompt: coherent Python `remove_duplicates` body, not garbage.
- Done event: `spec_path:"pp-mtp"`, `prefill_tok_s:999.8`, `decode_tok_s:121.3`, `tok_s:116.5`, `tau:3.52`, `accept_rate:2.52`, `cycles:27`, `pp:2`.

Related: #382

### CODE REVIEW

**Comment by unverbraucht** [comment]
## Hetero-pair validation: gfx906 (MI60) + gfx1031 (RX 6700 XT)

Validated this PR on a **mixed-arch** pair to de-risk the path that our fork (#382) was built around. Our hardware is a 32 GB CDNA1 card + a 12 GB RDNA2 card — a deliberately hostile setup for the uniform-split assumption and peer-access fast path.

### Setup

- `HIPFIRE_ALLOW_MIXED_ARCH=1 HIPFIRE_UNIFORM_VRAM_TOLERANCE_GB=99`
- Trunk: `qwen3.6-27b.mq4` (15 GB)
- MTP head: `qwen3.6-27b-cvs16384.mtp` (compressed, cvs=16384)
- Load params: `{"max_seq":2048,"pp":2,"kv_mode":"q8","mtp_mode":"on","mtp_k":3,"mtp_kv_mode":"q8"}`
- Prompt (byte-identical, committed-string, temp=0): sliding-window `longest_unique_substring`

### Results

Build is clean. Load reports the expected configuration:

```
MTP head loaded on output_device=1 (n_embd=5120, vocab=248320, cvs=16384, k=3, kv=Q8)
pp=2 loaded: layer_to_device=[0×32, 1×32], output_device=1, peer_access=true
```

Path is taken as intended — `spec_path:"pp-mtp"` on every run, no silent fallback to `generate_multi`.

3-run fresh-process median (after a warmup generate in each session):

| run | decode_tok_s | prefill_tok_s | tau | accept_rate | cycles |
|-----|--------------|---------------|------|-------------|--------|
| 1   | 16.2         | 126.5         | 2.77 | 1.77        | 92     |
| 2   | 16.1         | 126.4         | 2.77 | 1.77        | 92     |
| 3   | 16.1         | 126.5         | 2.77 | 1.77        | 92     |

**Greedy determinism holds:** decoded text md5 is byte-identical across all three runs (`a47c4186…`). Output is coherent — a structured thinking trace into a correct sliding-window implementation, no token attractors or repetition. No panics, no illegal-memory-access crashes, no VRAM leaks across the load/generate/unload cycle.

### The interesting part: Opt 3 isn't needed

Our abandoned #382 carried an optimization (per-band GDN tape-shard fast replay) specifically targeting PP+MTP partial-accept replay cost. On the **old** master, full-prefix replay gave us ~12.0 tok/s and the tape-shard replay lifted it to 14.6 tok/s (+21.7%).

On **this** branch's full-replay path, the same workload runs at **16.1 tok/s** — already *ahead* of our optimized path, with a worse split (uniform 32/32 vs our hand-tuned 52/12). The base master got fast enough that the tape-shard complexity no longer pays for itself. The "correctness first, skip the tape-shard replay" scope call in the PR description was the right one; we don't plan to resurrect Opt 3.

### One caveat for other testers (not a code defect)

A *non-compressed* MTP head is rejected cleanly rather than silently misused. We hit this with a `qwen3.5-9b.mtp` sidecar we had extracted with `mtp_extract` (no `compressed_vocab_size` baked in):

```
load failed: PP+MTP requires compressed-sidecar .mtp
```

Clean error, no panic. Worth flagging in a load-time error message maybe, but functionally correct.

### TL;DR

Mergeable as-is for the hetero/mixed-arch case. Deterministic, coherent, and faster than our fork's optimized variant. We'll abandon #382 in favor of this once it lands.
